### PR TITLE
Fix connReaper goroutine may leak

### LINKED ISSUE #153: The connReaper goroutine may leak
It seems the socket.Close() only does sck.reaperCond.Signal() to close connReaper goroutine. But the signal seems can be delivered only if the connReaper is in sck.reaperCond.Wait().

That cause at least short leaving sockets (.Dial() and .Close()) sometimes to leak.

### PR DESCRIPTION
**Pull request by egorse**
Fix #153

### CODE REVIEW

File: reaper/reaper_test.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
+// Copyright 2024 The go-zeromq Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package reaper
```
Comment (sbinet) [inline reaper/reaper_test.go:5]: wouldn't this work as well ? (and put it under the root of zmq4)
```suggestion
package zmq4_test
```

File: reaper/reaper_test.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/go-zeromq/zmq4"
+	"go.uber.org/goleak"
+)
+
+// TestReaper does multiple rapid Dial/Close to check that connection reaper goroutines are not leaking.
+// In is in own package as goleak detects also threads from objects instantiated by init().
```
Comment (sbinet) [inline reaper/reaper_test.go:19]:
```suggestion
// It is in a _test package as goleak detects also threads from values created during init().
```

File: reaper/reaper_test.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+func TestReaper(t *testing.T) {
+	defer goleak.VerifyNone(t)
+	//assert := asserts.New(t)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	rep := zmq4.NewRep(ctx)
+	ep := "ipc://@test.rep.socket"
+	err := rep.Listen(ep)
+	_ = err // assert.NoError(err)
```
Comment (sbinet) [inline reaper/reaper_test.go:28]:
```suggestion
	err := rep.Listen(ep)
	if err != nil { t.Fatal(err) }
```

File: reaper/reaper_test.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+	wgServer := &sync.WaitGroup{}
+	client := func() {
+		defer wgClients.Done()
+		for n := 0; n < maxMsgs; n++ {
+			func() {
+				ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+				defer cancel()
+				req := zmq4.NewReq(ctx)
+				err := req.Dial(ep)
+				_ = err // assert.NoError(err)
```
Comment (sbinet) [inline reaper/reaper_test.go:42]:
```suggestion
				if err != nil { t.Fatal(err) }
```

File: reaper/reaper_test.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+				err = req.Close()
+				_ = err // assert.NoError(err)
```
Comment (sbinet) [inline reaper/reaper_test.go:45]:
```suggestion
				if err != nil { t.Fatal(err) }
```

File: reaper/reaper_test.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+	}
+	server := func() {
+		defer wgServer.Done()
+		pong := zmq4.NewMsgString("pong")
+		for {
+			msg, err := rep.Recv()
+			if errors.Is(err, context.Canceled) {
+				break
+			}
+			// assert.NoError(err)
```
Comment (sbinet) [inline reaper/reaper_test.go:57]:
```suggestion
```

File: reaper/reaper_test.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+			if err != nil {
+				break
+			}
+			_ = msg // assert.Equal(string(msg.Frames[0]), "ping")
+			err = rep.Send(pong)
+			_ = err // assert.NoError(err)
```
Comment (sbinet) [inline reaper/reaper_test.go:63]:
```suggestion
			err = rep.Send(pong)
			if err != nil { t.Fatal(err) }
```

File: reaper/reaper_test.go (hunk 1 above)
Comment (egorse) [reply reaper/reaper_test.go:5]: No it wont works - the zmq4_test has init() which allocates sockets/goroutines, and goleak complains about those. So this test has to be fully excluded. atm I did it this way

Comment (egorse) [reply reaper/reaper_test.go]: Renamed to leaks_test
May be that would be better

File: leaks_test/reaper_leak_test.go [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/go-zeromq/zmq4"
+	"go.uber.org/goleak"
+)
+
+// TestReaper does multiple rapid Dial/Close to check that connection reaper goroutines are not leaking.
+// In is in own package as goleak detects also threads from values created during init().
```
Comment (sbinet) [inline leaks_test/reaper_leak_test.go:19]:
```suggestion
// TestReaper is in a dedicated package as goleak detects also goroutines from values created during init().
```

**Review by sbinet — APPROVED** [review]
LGTM once the minor cosmetic/doc change is applied.

thanks for the PR (and your relentless work on go-zeromq)

**Pull request merged**